underfs/gsg/src/main/java/alluxio/underfs/gsg/GSGOutputStream.java

```diff
@@ -0,0 +1,133 @@
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for writing a file into GSG. The data will be persisted to a temporary directory on the
+ * local disk and copied as a complete file when the {@link #close()} method is called.
+ */
+@NotThreadSafe
+public final class GSGOutputStream extends OutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(GSGOutputStream.class);
+
+  /** Bucket name of the Alluxio GSG bucket. */
+  private final String mBucketName;
+
+  /** Key of the file when it is uploaded to GSG. */
+  private final String mKey;
+
+  /** The Google cloud storage client. */
+  private final Storage mClient;
+
+  /** The pre-allocated buffer for single byte write operation. */
+  private final ByteBuffer mSingleByteBuffer;
+
+  /** The write channel of Google storage object. */
+  private WriteChannel mWriteChannel;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  private AtomicBoolean mClosed = new AtomicBoolean(false);
+
+  /**
+   * Constructs a new stream for writing a file.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param client the JetS3t client
+   */
+  public GSGOutputStream(String bucketName, String key, Storage client) throws IOException {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(), "Bucket name must "
+        + "not be null or empty.");
+    mBucketName = bucketName;
+    mKey = key;
+    mClient = client;
+    mSingleByteBuffer = ByteBuffer.allocate(1);
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    if (mWriteChannel == null) {
+      createWriteChannel();
+    }
+    mSingleByteBuffer.clear();
+    mSingleByteBuffer.putInt(b);
+    mWriteChannel.write(mSingleByteBuffer);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (mWriteChannel == null) {
+      createWriteChannel();
+    }
+    ByteBuffer buffer = ByteBuffer.wrap(b, off, len);
+    mWriteChannel.write(buffer);
+  }
+
+  @Override
+  public void flush() throws IOException {
+    // Google storage write channel do not support flush
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed.getAndSet(true)) {
+      return;
+    }
+    if (mWriteChannel == null) {
+      createWriteChannel();
```

Comment:
Do we need to create the channel in case its a 0 byte file?

Reply:
When 0 byte, we directly create an empty file. Thanks